mesa/experimental/cell_space.py

```diff
@@ -0,0 +1,247 @@
+            if not include_center:
+                neighborhood.pop(self, None)
+            return CellCollection(neighborhood)
+
+
+class CellCollection:
+    def __init__(self, cells: dict[Cell, list[Agent]] | Iterable[Cell]) -> None:
+        if isinstance(cells, dict):
+            self._cells = cells
+        else:
+            self._cells = {cell: cell.agents for cell in cells}
+
+    def __iter__(self):
+        return iter(self._cells)
+
+    def __getitem__(self, key):
+        return self._cells[key]
+
+    @cached_property
+    def __len__(self):
+        return len(self._cells)
+
+    def __repr__(self):
+        return f"CellCollection({self._cells})"
+
+    @cached_property
+    def cells(self):
+        return list(self._cells.keys())
+
+    @property
+    def agents(self):
+        return itertools.chain.from_iterable(self._cells.values())
+
+    def select_random_cell(self):
+        return random.choice(self.cells)
+
+    def select_random_agent(self):
+        return random.choice(list(self.agents))
+
+    def select(self, filter_func: Optional[Callable[[Cell], bool]] = None, n=0):
+        if filter_func is None and n == 0:
+            return self
+
+        return CellCollection(
+            {
+                cell: agents
+                for cell, agents in self._cells.items()
+                if filter_func is None or filter_func(cell)
+            }
+        )
+
+
+class Space:
+    cells: dict[Coordinate, Cell] = {}
+
+    def _connect_single_cell(self, cell):
+        ...
+
+    @cached_property
+    def all_cells(self):
+        return CellCollection({cell: cell.agents for cell in self.cells.values()})
+
+    def __iter__(self):
+        return iter(self.cells.values())
+
+    def __getitem__(self, key):
+        return self.cells[key]
+
+    def move_agent(self, agent: Agent, pos: Coordinate) -> None:
+        """Move an agent from its current position to a new position."""
+        self.cells[pos].add_agent(agent)
+
+    @property
+    def empties(self) -> CellCollection:
+        return self.all_cells.select(lambda cell: cell.is_empty)
+
+    def move_to_empty(self, agent: Agent) -> None:
+        while True:
+            new_cell = self.all_cells.select_random_cell()
+            if new_cell.is_empty:
+                new_cell.add_agent(agent)
+                return
+
+        # TODO: Adjust cutoff value for performance
+        for _ in range(len(self.all_cells) // 10):
+            new_cell = self.all_cells.select_random_cell()
+            if new_cell.is_empty:
+                new_cell.add_agent(agent)
+                return
+
+        try:
+            self.empties.select_random_cell().add_agent(agent)
+        except IndexError as err:
+            raise Exception("ERROR: No empty cell found") from err
+
+
+class Grid(Space):
+    def __init__(
+        self, width: int, height: int, torus: bool = False, moore: bool = True, capacity=1
+    ) -> None:
```

Comment:
capacity is now very nicely done!. No need for MultiGrid and SingleGrid anymore.